src/core/io_queue.cc

```diff
@@ -404,26 +404,27 @@ std::vector<io_request::part> io_request::split(size_t max_length) {
 
 std::vector<io_request::part> io_request::split_buffer(size_t max_length) {
     std::vector<part> ret;
-    ret.reserve((_size.len + max_length - 1) / max_length);
+    const auto& op = _read;
```

Comment:
This caught my eye. The `split_buffer` is called for `operation::read` and `operation::write` and this place relies on three assumptions:

0. `read_op _read` and `write_op _write` are unioned. Not that it's going to change ever, but this statement is needed for correctness
1. `using write_op = read_op` exists. This is worse, because compiler won't warn us if some time it changes, e.g. -- zns patchset extends write_op to carry append data on board
2. given 1 is true, compiler lays out `read_op _read` and `write_op _write` matching each other bit-to-bit AND accessing `_read` after initialization took place over `_write` is OK. I don't know standard/compiler well enough to judge how OK is that

To sum up -- item 2 is of the greatest concern on my side

Reply:
@xemul agreed. i agree that the 2nd assumption is fragile. so a `static_assert()` should do help us make this explicit and fail the build if the assumption does not hold anymore.